File: ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/db/AemImportRepository.java
package org.cru.godtools.articles.aem.db;

import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Transaction;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.WorkerThread;

import org.cru.godtools.articles.aem.model.AemImport;
import org.cru.godtools.articles.aem.model.Article;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

@Dao
public abstract class AemImportRepository {
    @NonNull
    private final ArticleRoomDatabase mDb;

    AemImportRepository(@NonNull final ArticleRoomDatabase db) {
        mDb = db;
    }

    @Transaction
    @WorkerThread
    public void processAemImportSync(@NonNull final AemImport aemImport, @NonNull final List<Article> articles) {
        final ArticleDao articleDao = mDb.articleDao();
        final AemImportDao aemImportDao = mDb.aemImportDao();

        final List<Uri> articleUris = new ArrayList<>();
        for (final Article article : articles) {
            articleUris.add(article.uri);

            // insert/update article
            articleDao.insertOrIgnore(article);
            articleDao.update(article.uri, article.uuid, article.title);

            // associate the article with this AemImport
            aemImportDao.insertOrIgnore(new AemImport.AemImportArticle(aemImport, article));

            // replace all categories
            articleDao.removeAllCategories(article.uri);
            articleDao.insertOrIgnore(article.getCategoryObjects());
        }

        // remove any articles that are no longer part of the AemImport
        aemImportDao.removeOldArticles(aemImport.uri, articleUris);

        // update the last processed time
        aemImportDao.updateLastProcessed(aemImport.uri, new Date());
    }
}

File: ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/db/AemImportRepository.kt
package org.cru.godtools.articles.aem.db

import androidx.annotation.WorkerThread
import androidx.room.Dao
import androidx.room.Transaction
import org.cru.godtools.articles.aem.model.AemImport
import org.cru.godtools.articles.aem.model.Article
import java.util.Date

@Dao
abstract class AemImportRepository internal constructor(private val db: ArticleRoomDatabase) {
    @Transaction
    @WorkerThread
    open fun processAemImportSync(aemImport: AemImport, articles: List<Article>) {
        // insert/update any supplied articles
        with(db.articleDao()) {
            articles.forEach {
                insertOrIgnore(it)
                update(it.uri, it.uuid, it.title)

                // replace all tags
                removeAllTags(it.uri)
                insertOrIgnoreTags(it.tagObjects)
            }
        }

        // update associations between the AemImport and the articles
        with(db.aemImportDao()) {
            insertOrIgnoreArticles(articles.map { AemImport.AemImportArticle(aemImport, it) })
            removeOldArticles(aemImport.uri, articles.map { it.uri })

            // update the last processed time
            updateLastProcessed(aemImport.uri, Date())
        }

        // remove any orphaned articles
        db.articleRepository().removeOrphanedArticles()
    }

    @Transaction
    @WorkerThread
    open fun removeOrphanedAemImports() {
        db.aemImportDao().removeOrphanedAemImports()
        db.articleRepository().removeOrphanedArticles()
    }
}
